### PR TITLE
Creating and Uploading dumps for hung processes in a single invocation on Helix machines

### PR DESCRIPTION
**Pull request by Anipik**
In this we are creating folders for the specific hung processes. Earlier we used to have two calls to dumpling, First one to create dumps and second one to upload it. But now there is new api to do both works in a single invocation

### CODE REVIEW

**Comment by Anipik** [comment]
@danmosemsft  @brianrob can you review the changes

**Comment by danmosemsft** [comment]
@Anipik this title is not sufficient for folks not involved in the work. Please update the change description (top comment and one of the commits) to explain the purpose of the changes.

File: src/web/Content/client/dumpling.py [hunk 1]
```diff
@@ -880,16 +879,46 @@ def _download_dump(self, dir, dumpManifest):
...
+            self.Hang(config)
+
     def Hang(self, config):
-        if os.path.exists(config.dbgpath) and os.path.isdir(config.outpath):
-            CommandProcessor._create_hang_dump(config.pid, config.outpath, config.dbgpath)
+        if os.path.exists(config.dbgpath) and os.path.isdir(config.outpath):            
             process = psutil.Process(int(config.pid))
+            parent_dump_folder = os.path.join(config.outpath, "process" + str(config.pid))
+
+            if not os.path.exists(parent_dump_folder):
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: rewrite as
```python
+            if os.path.exists(parent_dump_folder):
 +                shutil.rmtree(parent_dump_folder)

 +                os.makedirs(parent_dump_folder)
```

File: src/web/Content/client/dumpling.py [hunk 2]
```diff
@@ -880,16 +879,46 @@ def _download_dump(self, dir, dumpManifest):
...
+
+    def HangAndUpload(self, config):
+        self.Hang(config)
+        parent_dump_folder = os.path.join(config.outpath, "process" + str(config.pid))
+        
+        if config.incpaths is None:
+            config.incpaths = [parent_dump_folder]
+        
+        if config.dumppath is None:
+            dump_file_name = "memdump" + str(config.pid) + ".dmp"
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: can we include the process names in the dump file names? eg so I can distinguish `memdump.xunit.console.1234.dmp` from `memdump.ping.2345.dmp` and pick the one I wan tto open?

File: src/web/Content/client/dumpling.py [hunk 3]
```diff
@@ -880,16 +879,46 @@ def _download_dump(self, dir, dumpManifest):
...
+            config.dumppath = os.path.join(parent_dump_folder, dump_file_name)
+        
+        if config.displayname is None:
+            config.displayname = "HungProcess" + str(config.pid);
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: Hopefully we normally pass in a better name, that includes the test library name?

File: src/web/Content/client/dumpling.py [hunk 4]
```diff
@@ -1119,14 +1148,30 @@ def _parse_args(argv):
...
+
+    hung_parser.add_argument('--upload', type=bool, default=False, help='the path to the directory for memory dump file')
+
+    hung_parser.add_argument('--user', type=str, default=getpass.getuser().lower(), help='The username to pass to the dumpling service.  This argument is ignored unless --dumppath is specified')     
     
+    hung_parser.add_argument('--triage', choices=['none', 'client', 'full'], default='client', help='specifies the triage info to be uploadeded with the dump')
+
+    hung_parser.add_argument('--properties', nargs='*', type=_parse_key_value_pair, help='a list of properties to be associated with the dump in the format key=value', metavar='key=value')  
+                                         
+    hung_parser.add_argument('--propfile', type=argparse.FileType('r'), help='path to a file containing a json serialized dictionary of property value paires')
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: typo pairs

File: src/web/Content/client/dumpling.py [hunk 5]
```diff
@@ -1119,14 +1148,30 @@ def _parse_args(argv):
...
+    hung_parser.add_argument('--dbgpath', type=str, required=True, help='path to debugger to be used by the dumpling client for creating dump')
                                                  
-    debug_parser.add_argument('--outpath', type=str, default=os.getcwd(), help='the path to the directory for memory dump file')     
+    hung_parser.add_argument('--outpath', type=str, default=os.getcwd(), help='the path to the directory for memory dump file')
+
+    hung_parser.add_argument('--upload', type=bool, default=False, help='the path to the directory for memory dump file')
+
+    hung_parser.add_argument('--user', type=str, default=getpass.getuser().lower(), help='The username to pass to the dumpling service.  This argument is ignored unless --dumppath is specified')     
     
+    hung_parser.add_argument('--triage', choices=['none', 'client', 'full'], default='client', help='specifies the triage info to be uploadeded with the dump')
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: typo

File: src/web/Content/client/dumpling.py [hunk 6]
```diff
@@ -962,9 +991,9 @@ def _create_hang_dump(pid, outpath, debuggerpath):
         osStr = platform.system().lower()
         command = ""
         if osStr == 'linux':
-            command = "./" + debuggerpath + " " + pid + " --name " + outpath + "\memdum" + pid + ".dmp"
+            command = "./" + debuggerpath + " " + pid + " --name " + outpath + "\memdump" + pid + ".dmp"
         elif osStr == 'windows':
-            outputpath = outpath + "\memdum" + pid + ".dmp"
```
Comment (danmosemsft) [inline src/web/Content/client/dumpling.py]: was anything reading from `memdum` before? If so does it need a rename also

Comment (Anipik) [reply src/web/Content/client/dumpling.py]: Earlier it was used in helix but now we have shifted this part to dumpling only

**Comment by DrewScoggins** [comment]
Why are we making this change and adding the upload inside of the hang command?

**Comment by Anipik** [comment]
Matt suggested as we are not using it to create dumps anywhere else. so its better to combine them in a single invocation. @MattGal

**Comment by DrewScoggins** [comment]
But we already have an upload verb for dumpling, and now we have to copy some, but not all of the command line arguments for upload into the hang verb. This means that if we ever have to make changes to one we have to move the changes around. It seems better to me to simply call the upload verb we already have and not recreate the functionality inside hang.

**Comment by Anipik** [comment]
yeah you are right about that. @MattGal should we just keep them into two separate hang and upload modules ?

**Comment by Anipik** [comment]
@DrewScoggins  I have removed the upload from Hang..can you take a look

**Comment by DrewScoggins** [comment]
So this change as it stands is just to create a dump folder to drop the hang dumps in, yes?

**Comment by Anipik** [comment]
yeah and just changed the name of the dump files to include the process names

**Comment by DrewScoggins** [comment]
LGTM

**Comment by DrewScoggins** [comment]
Can you rebase this down to logical commits and have good commit messages for each one that describe what is going on?

After you do that I will commit this.

**Comment by Anipik** [comment]
@DrewScoggins  I have squashed them to logical commits.

**Comment by Anipik** [comment]
The final Commit message for master branch can be "Dumps getting created in appropriate\respective folders"

**Pull request merged**